In [55]:
import pandas as pd
from bs4 import SoupStrainer
from bs4 import BeautifulSoup as bs
import requests

Experiment with one phone:

In [71]:
r = requests.get("https://www.pickaboo.com/realme-narzo-50a-prime-4gb-128gb.html")

In [72]:
soup = bs(r.content)

soup.prettify()
print(soup)

<!DOCTYPE html>
<html lang="en">
<head prefix="og: http://ogp.me/ns# fb: http://ogp.me/ns/fb# product: http://ogp.me/ns/product#">
<script>
    var BASE_URL = 'https://www.pickaboo.com/';
    var require = {
        "baseUrl": "https://www.pickaboo.com/pub/static/frontend/Emthemes/everything_hitech/en_US"
    };
</script>
<meta charset="utf-8"/>
<meta content="realme Narzo 50a Prime 4GB/128GB realme Narzo 50a Prime 4GB/128GB

Operating System: realme UI R Edition | Based on Android 11
Processor: Unisoc T612 Processor
CPU: Octa-core, 12nm, up to 1.82GHz
GPU: ARM Mali-G57
RAM: 4GB LPDDR4X
ROM: 128GB UFS 2." name="description"/>
<meta content="realme Narzo 50a Prime 4GB/128GB" name="keywords"/>
<meta content="INDEX,FOLLOW" name="robots"/>
<meta content="realme Narzo 50a Prime 4GB/128GB" name="title"/>
<meta content="width=device-width, initial-scale=1" name="viewport"/>
<meta content="telephone=no" name="format-detection"/>
<title>realme Narzo 50a Prime 4GB/128GB</title>
<link href="https

In [73]:
product_attibute = soup.find(class_ = "data table additional-attributes")
print(product_attibute.prettify())

<table class="data table additional-attributes" id="product-attribute-specs-table">
 <caption class="table-caption">
  More Information
 </caption>
 <tbody>
  <tr>
   <th class="col label" scope="row">
    Brand
   </th>
   <td class="col data" data-th="Brand">
    Realme
   </td>
  </tr>
  <tr>
   <th class="col label" scope="row">
    SKU
   </th>
   <td class="col data" data-th="SKU">
    RN50AP4128
   </td>
  </tr>
  <tr>
   <th class="col label" scope="row">
    Express Delivery
   </th>
   <td class="col data" data-th="Express Delivery">
    No
   </td>
  </tr>
  <tr>
   <th class="col label" scope="row">
    Warranty Information
   </th>
   <td class="col data" data-th="Warranty Information">
    12 Months Official Warranty
   </td>
  </tr>
  <tr>
   <th class="col label" scope="row">
    Emi Available
   </th>
   <td class="col data" data-th="Emi Available">
    0
   </td>
  </tr>
  <tr>
   <th class="col label" scope="row">
    Display Size (Inches)
   </th>
   <td class="col 

In [74]:
product_rows = product_attibute.find_all("tr")
for i in product_rows:
    print(i.prettify())
    print("+++===========================+++==========================+++======================+++")

<tr>
 <th class="col label" scope="row">
  Brand
 </th>
 <td class="col data" data-th="Brand">
  Realme
 </td>
</tr>

+++===========================+++==========================+++======================+++
<tr>
 <th class="col label" scope="row">
  SKU
 </th>
 <td class="col data" data-th="SKU">
  RN50AP4128
 </td>
</tr>

+++===========================+++==========================+++======================+++
<tr>
 <th class="col label" scope="row">
  Express Delivery
 </th>
 <td class="col data" data-th="Express Delivery">
  No
 </td>
</tr>

+++===========================+++==========================+++======================+++
<tr>
 <th class="col label" scope="row">
  Warranty Information
 </th>
 <td class="col data" data-th="Warranty Information">
  12 Months Official Warranty
 </td>
</tr>

+++===========================+++==========================+++======================+++
<tr>
 <th class="col label" scope="row">
  Emi Available
 </th>
 <td class="col data" data-th="Emi Availabl

In [75]:
phone_info = {}
for i in product_rows:
    content_key = i.find("th").get_text()
    content_value = i.find("td").get_text()
    phone_info[content_key] = content_value

In [76]:
print(phone_info)

{'Brand': 'Realme', 'SKU': 'RN50AP4128', 'Express Delivery': 'No', 'Warranty Information': '12 Months Official Warranty', 'Emi Available': '0', 'Display Size (Inches)': '6.6', 'Front Camera': '8 MP', 'RAM(GB)': '4', 'SIM Card': 'Dual Nano SIM', 'Processor': 'Unisoc T612', 'CPU Speed(Ghz)': 'up to 1.82GHz', 'Connectivity': 'Wi-Fi: 2.4/5GHz | Bluetooth 5.0', 'Battery mAh': '5000 mAh', 'Back Camera': '50MP AI Triple Camera', 'Phone Model': 'Narzo 50a Prime'}


Now get similar data for every phone:

1. extract a tags and retrieve href links from it. 
2. Save it as a list

In [217]:
# link_ = "https://www.pickaboo.com/smartphone.html?p=1&product_list_limit=80&product_list_order=price&product_list_dir=desc"

link_ = "https://www.pickaboo.com/smartphone.html?p=2&product_list_limit=80&product_list_order=price&product_list_dir=desc"

r_link = requests.get(link_)
soup_link = bs(r_link.content)

In [218]:
tag_links = soup_link.find_all(class_ = "product name product-item-name")

Get Spec information for all the phones in Pickaboo site

In [206]:
#function to get phone information.
def get_phone_info(link):
    soup = bs(requests.get(link).content)

    product_attibute = soup.find(class_ = "data table additional-attributes")
    product_rows = product_attibute.find_all("tr")

    phone_info = {}
    for i in product_rows:
        content_key = i.find("th").get_text()
        content_value = i.find("td").get_text()
        phone_info[content_key] = content_value
    
    return phone_info

In [207]:
list_of_phone_info = []

Saving all the dictionary values in a list

In [219]:

for i in range(len(tag_links)):
    try:
        individual_link = tag_links[i].find(class_ = "product-item-link")['href']
        list_of_phone_info.append(get_phone_info(individual_link))
    except AttributeError as a:
        pass

In [220]:
list_of_phone_info

[{'Brand': 'Samsung',
  'SKU': 'SGZF35G',
  'Offers': 'Checkout + 0% EMI',
  'Express Delivery': 'No',
  'Warranty Information': '12 Months Official Warranty',
  'Emi Available': '18 Months',
  'Front Camera': '4 MP',
  'Cover Camera': '10MP',
  'System Storage ': '256 GB',
  'RAM(GB)': '12GB',
  'SIM Card': 'Nano + eSIM',
  'Processor': 'Qualcomm Snapdragon 888',
  'CPU Speed(Ghz)': '5nm 64-bit Octa-Core Processor ※ 2.84GHz (Maximum Clock Speed) + 2.4 GHz + 1.8 GHz',
  'Connectivity': 'Wi-Fi 802.11 a/b/g/n/ac/ax, HE80, MIMO, 1024-QAM. •\tBluetooth® v 5.0 (LE up to 2Mbps), USB type-C, NFC, Location (GPS, Galileo, Glonass, BeiDou)',
  'Battery mAh': '4400mAh (typical) dual battery.  Fast Charging 25W',
  'Back Camera': '12MP+12MP+12MP',
  'Phone Model': 'Samsung Galaxy Z Fold3 5G'},
 {'Brand': 'Samsung',
  'SKU': 'SMSNGULTRA',
  'Offers': 'Checkout + 0% EMI',
  'Express Delivery': 'No',
  'Warranty Information': '12 Months Official Warranty',
  'Emi Available': '18 Months'},
 {'Brand': 

Saving all the dictionary information in JSON file

In [221]:
import json

def save_data(title, data):
    with open(title, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

In [222]:
import json

def load_data(title):
    with open(title, 'w', encoding='utf-8') as f:
        return json.load(f)

In [223]:
save_data("list_of_phone_info.json", list_of_phone_info)

In [226]:
df = pd.DataFrame(list_of_phone_info)

In [227]:
df.head()

Brand            SKU             Offers Express Delivery  \
0  Samsung        SGZF35G  Checkout + 0% EMI               No   
1  Samsung     SMSNGULTRA  Checkout + 0% EMI               No   
2  Samsung     SMSHNGGPLS  Checkout + 0% EMI               No   
3   Xiaomi  XMI12P12256GB   Checkout+ 0% EMI               No   
4   Xiaomi   XMI12P8256GB   Checkout+ 0% EMI               No   

          Warranty Information Emi Available Front Camera Cover Camera  \
0  12 Months Official Warranty     18 Months         4 MP         10MP   
1  12 Months Official Warranty     18 Months          NaN          NaN   
2  12 Months Official Warranty     12 Months         10MP          NaN   
3  12 Months Official Warranty     12 Months        32 MP          NaN   
4  12 Months Official Warranty     12 Months        32 MP          NaN   

  System Storage  RAM(GB)  ...                    Processor  \
0          256 GB    12GB  ...      Qualcomm Snapdragon 888   
1             NaN     NaN  ...                          NaN   
2          256 GB       8  ...                          NaN   
3          256 GB    12GB  ...  Qualcomm Snapdragon 8 Gen 1   
4          256 GB       8  ...  Qualcomm Snapdragon 8 Gen 1   

                                      CPU Speed(Ghz)  \
0  5nm 64-bit Octa-Core Processor ※ 2.84GHz (Maxi...   
1                                                NaN   
2                                                NaN   
3      Up to 3.0GHz, with Armv9 Cortex-X2 technology   
4      Up to 3.0GHz, with Armv9 Cortex-X2 technology   

                                        Connectivity  \
0  Wi-Fi 802.11 a/b/g/n/ac/ax, HE80, MIMO, 1024-Q...   
1                                                NaN   
2                                                NaN   
3     2.4GHz/5GHz Wi-Fi | Bluetooth: 5.2 | GPS | NFC   
4     2.4GHz/5GHz Wi-Fi | Bluetooth: 5.2 | GPS | NFC   

                                         Battery mAh     Back Camera  \
0  4400mAh (typical) dual battery.  Fast Charging...  12MP+12MP+12MP   
1                                                NaN             NaN   
2                                            4500mAh             NaN   
3  4600mAh (typ) battery | 120W Xiaomi HyperCharg...             NaN   
4  4600mAh (typ) battery | 120W Xiaomi HyperCharg...             NaN   

                 Phone Model Display Size (Inches)  \
0  Samsung Galaxy Z Fold3 5G                   NaN   
1                        NaN                   NaN   
2                        NaN                   6.6   
3             Xiaomi 12 Pro                    NaN   
4             Xiaomi 12 Pro                    NaN   

                                     Extra Features  Delivery Information  \
0                                                NaN                  NaN   
1                                                NaN                  NaN   
2                                                NaN                  NaN   
3  LiquidCool Technology (Super big VC + 3 graphi...                  NaN   
4  LiquidCool Technology (Super big VC + 3 graphi...                  NaN   

  Coming Soon  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 21 columns]

In [228]:
df.shape

(191, 21)

In [229]:
df.to_csv("list_of_phone_info_bs4.csv")

Let's add the price 

In [234]:
# price_link = "https://www.pickaboo.com/smartphone.html?p=1&product_list_limit=80&product_list_order=price&product_list_dir=desc"
price_link = "https://www.pickaboo.com/smartphone.html?p=2&product_list_limit=80&product_list_order=price&product_list_dir=desc"

soup_price = bs(requests.get(price_link).content)

In [231]:
list_of_phone_price = []

In [235]:

product_price = soup_price.find_all(class_ = "special-price")

for i in product_price:
    list_of_phone_price.append(i.find(class_ = "price").get_text().replace("৳", "").replace(",", ""))
    
print(list_of_phone_price)

['195999', '164999', '127999', '89999', '79999', '76990', '76699', '69999', '66990', '58990', '52999', '52499', '49999', '45999', '42999', '42990', '41699', '39999', '39999', '39990', '39990', '38999', '37990', '35599', '32990', '32990', '31990', '31990', '31599', '29999', '28999', '27999', '27990', '27990', '26999', '25990', '23999', '23999', '23499', '23499', '23399', '22999', '22999', '22990', '21499', '20990', '19999', '20599', '19999', '19999', '19990', '19299', '18999', '18999', '18990', '17990', '17499', '16990', '16490', '16299', '15999', '15490', '14999', '14999', '14990', '14990', '14990', '14990', '14599', '13999', '13999', '13999', '13999', '13500', '12999', '12499', '11999', '9999', '3799', '195999', '164999', '127999', '89999', '79999', '76990', '76699', '69999', '66990', '58990', '52999', '52499', '49999', '45999', '42999', '42990', '41699', '39999', '39999', '39990', '39990', '38999', '37990', '35599', '32990', '32990', '31990', '31990', '31599', '29999', '28999', '2799

In [ ]:
len(list_of_phone_price)

Adding some blank elements to the list to match the Length of values with DataFrame's length of index (191) as some of the phones doens't have price information in Pickaboo Site.

In [240]:
for i in range(33):
    list_of_phone_price.append("")

In [245]:
df['Price'] = list_of_phone_price

In [246]:
df.to_csv("list_of_phone_info_bs4.csv")